### PR TITLE
feature/s21_strchr 

### LINKED ISSUE #25: `s21_strchr` - Finds the first occurrence of `c` in `str`.
## Description
`int strncmp(const char *str1, const char *str2, size_t n)`
Implement the `s21_strncmp()` function to compare at most the first `n` bytes of strings `str1` and `str2`, matching the behavior of the standard `strncmp()` function (POSIX.1-2017).

**Requirements**
- [ ] Follow Google C++ Style Guide  
- [ ] 100% test coverage (via Check library)  
- [ ] No memory leaks  
- [ ] Documentation in header file  
- [ ] Match standard library behavior exactly including:
   - Return values (0, positive, negative)  
   - Comparison of partial strings  
   - Handling of null terminators  
   - Edge cases with n=0  

## Test Cases Needed
- Basic string comparison
- Partial comparison (n < string length)
- Equal strings
- Different length strings
- Empty strings
- Unicode/ASCII compatibility
- Comparison with n=0
- Early null terminator cases

## Files to Modify
- `src/s21_strncmp.c` (implementation)
- `src/s21_string.h` (declaration)
- `tests/test_strncmp.c` (unit tests)
- `tests/suite.h` (definition of `s21_strncmp_suite()`)
- `tests/main.c` (running of suites and new one)

### PR DESCRIPTION
**Pull request by ParthenGithub**
Closes #22

Function takes string and character and returns pointer to first character appearance in this string.

### CODE REVIEW

File: tests/main.c [hunk 1]
```diff
@@ -7,9 +7,10 @@
 
 int main(void) {
   int number_failed;
-  Suite *s = s21_strerror_suite();
+  Suite *s = s21_strchr_suite();
```
Comment (DarkDomian) [inline tests/main.c]: ## Why did you change that?

And after - you call the function `srunner_create(s)` and put into it your suite. Okay, but later you add same suite into srunner again with `srunnre_add_suite(sr, s21_strchr_suite())`

And more - now it deal error from `cppchekc` 

```bash
tests/test_strerror.c:36:0: style: The function 's21_strerror_suite' is never used. [unusedFunction]
Suite *s21_strerror_suite(void) {
^
```

Or turn that change around and restore suite runner initialization with `s21_strerror_suite()` or put that function back to suit later with `srunner_add_suite()`

File: headers/s21_string.h [hunk 2]
```diff
@@ -29,6 +30,13 @@ int s21_memcmp(const void *str1, const void *str2, s21_size_t n);
 void *s21_memcpy(void *dest, const void *src, s21_size_t n);
 void *s21_memset(void *str, int c, s21_size_t n);
 char *s21_strncat(char *dest, const char *src, s21_size_t n);
+/** 
+ * @brief Finds the first occurrence of c in str
+ * @author Evgeniy Parfenyuk (Parthen/rhydonte)
+ * @date June 19, 2025
+ * @version 1.0
+ * @return char *
```
Comment (DarkDomian) [inline headers/s21_string.h]: There is no error - just returned valuse as `char *` doesn't describe clear the function work.

I take that such information from `man pages`
For that function I like to write:

```
/**
 * @brief locate character in string
 * @return the `s21_strchr()` function return a pointer to the matched character or NULL if  the  character is  not  found.
 * The terminating null byte is considered part of the string, so that if `c` is specified as `'\0'`,
 * these function return a pointer to the terminator.
 */
```

I just take all that from man. I don't know how good or bad that, but it much clearly in my mind.

What you think?

File: src/s21_strchr.c [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#include <string.h>
+// TODO: remove string.h and replace strlen when s21_strlen will be implemented
+#include "../headers/s21_string.h"
+
+// Why we should return char *, but take int?
+char *s21_strchr(const char *str, int c) {
```
Comment (DarkDomian) [inline src/s21_strchr.c:5]: `int` can take the `EOF` value, char cant

need to test `strchar()` work on string with `EOF` on the end

> Probably need to create file with text and use it into testing. I love to add `Alice's adventures in wonderland` chapter 1. HAha

File: tests/test_strchr.c
Comment (DarkDomian) [inline tests/test_strchr.c:1]: As I said, you need to check the work with strings with `EOF` on the end if possible

File: src/s21_strchr.c [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    if (*str == (char)c) {
+      return (char *)str;  // Cast to remove const
+    }
+    str++;
+  }
+  // Check the null terminator itself
+  if ((char)c == '\0') {
+    return (char *)str;
+  }
+  return NULL;  // Character not found
```
Comment (DarkDomian) [inline src/s21_strchr.c]: WOW... `NULL` is prohibited. Use `S21_NULL` instead

**Review by DarkDomian — CHANGES_REQUESTED** [review]
Want work on prohibited functions, and check `strchr()` work with `EOF`. Also suggestions about `doxygen` comments.

File: src/s21_strchr.c (hunk 4 above)
Comment (ParthenGithub) [reply src/s21_strchr.c]: It doesn't pass tests with S21_NULL (check.h thinks S21_NULL != NULL)

File: headers/s21_string.h (hunk 2 above)
Comment (ParthenGithub) [reply headers/s21_string.h]: Probably a good idea

Comment (DarkDomian) [reply headers/s21_string.h]: \> doesn't describe clear the function work.

When we use code browsering, we can get the functions signature - and we will see that function really return `char *`. And looking for comment about returning value - we wait a bit more informative guide fore us.

File: tests/main.c (hunk 1 above)
Comment (ParthenGithub) [reply tests/main.c]: I should dive deeper here, so ToDo later

File: src/s21_strchr.c (hunk 3 above)
Comment (DarkDomian) [reply src/s21_strchr.c:5]: Instead of making FILE, you can try `int buffer[] = {'t', 'e', 's', 't', EOF};` first

File: src/s21_strchr.c (hunk 4 above)
Comment (DarkDomian) [reply src/s21_strchr.c]: You function should return `S21_NULL`, just keep in mind on testing

File: headers/s21_string.h (hunk 2 above)
Comment (DarkDomian) [reply headers/s21_string.h]: Of course don't forget change NULL to S21_NULL, and functions name with prefix. Also change name of variable:
for example if original function signature work with variable `c` (from character)
but our function for same purpose use `chr` - change it into text of comment too

File: src/s21_strchr.c (hunk 3 above)
Comment (ParthenGithub) [reply src/s21_strchr.c:5]: >`int buffer[] = {'t', 'e', 's', 't', EOF};`

How I would give this to the function?)

File: tests/test_strchr.c [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  ck_assert_ptr_eq(result_s21, result_std);  // Compare pointers
+  if (result_s21 != NULL && result_std != NULL) {
+    ck_assert_str_eq(result_s21, result_std);  // Compare strings (both "\0")
+  }
+}
+END_TEST
+
+START_TEST(test_strchr_search_eof) {
+  const char *str = "I'm a little Bee and my name is Vitalya";
+  int c = -1;  // EOF
```
Comment (gashturik) [inline tests/test_strchr.c]: there  is an error of cppcheck `tests/test_strchr.c:54:32: error: Invalid strchr() argument nr 2. The value is -1 but the valid values are '0:255'. [invalidFunctionArg]
                   strchr(str, c));  // Compare pointers (both NULL)`
I included stdio.h and changed strings 53-55 to 
```c
// int c = -1;  // EOF
ck_assert_ptr_eq(s21_strchr(str, EOF),
                   strchr(str, EOF));  // Compare pointers (both NULL)
```

and it's working now (no cppcheck errors).

**Review by gashturik — APPROVED** [review]

**Review by DarkDomian — APPROVED** [review]
Need to consider about EOF check. Cos it's really interesting question why that function takes `int` instead `char`

File: tests/test_strchr.c (hunk 5 above)
Comment (ParthenGithub) [reply tests/test_strchr.c]: Fixed
